local_planner/src/nodes/local_planner_node.cpp

```diff
@@ -179,10 +178,8 @@ size_t LocalPlannerNode::numReceivedClouds() {
 size_t LocalPlannerNode::numTransformedClouds() {
   size_t num_transformed_clouds = 0;
   for (size_t i = 0; i < cameras_.size(); i++) {
-    std::unique_lock<std::mutex> lk(*(cameras_[i].trans_ready_mutex_));
-    cameras_[i].trans_ready_cv_->wait_for(
-        lk, std::chrono::milliseconds(30),
-        [this, i] { return cameras_[i].transformed_; });
+    std::lock_guard<std::mutex> transformed_cloud_guard(
+        *(cameras_[i].transformed_cloud_mutex_));
```

Comment:
Since localplanner_node running period is 100ms. https://github.com/PX4/avoidance/blob/master/local_planner/src/nodes/local_planner_node.cpp#L21

Just checking transform status(don't wait for transform finish) may cause don't have chance to run planner. Then the interval of planning will be longer than 100ms.

Reply:
this function is called from the main thread which also does all the queries on the rostopics (e.g. getting new pointclouds) if we block that until it is transformed we will not get any new pointclouds anymore. So we would try to transform a specific pointcloud until done, instead of trying to transform the most recent one if it takes too long.
Blocking at that point could even lead to a deadlock if the transformation of a specific pointcloud is impossible (dropped data, ...).

Reply:
The new pointcouds is triggered by callback, won't block by this function.
And it won't causes deadlock, since there is a max wait threshold time.

Reply:
The callback for the pointcloud is queried in the same thread, so if we wait at that location we will not get new pointclouds during the waiting period.
You are right, the max time works for one camera, but if we have multiple, we would also block for potentially x-times that timespan. Therefore, I would rather leave it that way - spinning around once more if the pointcloud is not yet transformed.

Reply:
I still not understand why it will causes cant get new cloudpoints issue. Waiting time out happen for multiple cameras,that's impossible,since the camera is 30fps.

And spinning once more will increase 100ms for interval of running planner.

Reply:
To why we do not get clouds in this time interval:

every 100ms `cmdLoopCallback` is executed. At the beginning of that function all the callbacks are evaluated: https://github.com/PX4/avoidance/blob/aeed406574119fb43a09958c82d28e0f52fc287b/local_planner/src/nodes/local_planner_node.cpp#L290

Next comes `updatePlanner()` which contains the waiting. Only when the waiting is finished and `cmdLoopCallback` is executed again, the new pointclouds are received

Reply:
We could technically wait those 30ms, that would be fine, but as soon as we use 3 cameras we would be at a max timeout of 90ms, as soon as that time is close to 100ms or higher we will slow down the main loop

Reply:
...the max timeout 90ms may happen, however, the possibility is very small... since there are 3 threads to handle each camera independently.
The callback trigger at the beginning of `cmdLoopCallback` may cause the latest cloud points not use in planner. Since cmdLoopCallback to updatePlanner() is very fast. Callback send the new point cloud then transform the point cloud need more time. The lastest point cloud may not be used in updatePlanner().
Anyway, i didn't test it yet....

Reply:
But since the camera frame rate is higher than the planner, and the transform is very fast, this way the planner always has the latest transformed point cloud available. It just gets whatever is there, even if maybe another one will be available soon. Otherwise the planner is delayed.